federation/pkg/kubefed/init/init.go

```diff
@@ -173,6 +176,7 @@ func (o *initFederationOptions) Bind(flags *pflag.FlagSet, defaultServerImage, d
 	flags.Int32Var(&o.apiServerNodePortPort, apiserverPortFlag, 0, "Preferred port to use for api server nodeport service (0 for random port assignment). Valid only if '"+apiserverServiceTypeFlag+"=NodePort'.")
 	flags.BoolVar(&o.apiServerEnableHTTPBasicAuth, "apiserver-enable-basic-auth", false, "Enables HTTP Basic authentication for the federation-apiserver. Defaults to false.")
 	flags.BoolVar(&o.apiServerEnableTokenAuth, "apiserver-enable-token-auth", false, "Enables token authentication for the federation-apiserver. Defaults to false.")
+	flags.StringVar(&o.tolerationsString, "tolerations", "", "If specified, the pod's tolerations. Example \"key1=value1:NoSchedule,key2=value2:NoSchedule\"")
```

Comment:
IIUC, this is about the federation api server?
The intent is to have federation-apiserver deployment tolerate some taints?
Why don't we just encode some tolerations into the API server's deployment instead of exposing these as user visible options?
My concern is that it is so easy to add options that people sometime ignore the difficulties to deprecate/remove them in the future.

Reply:
I agree with @tengqm here. Adding flags to customize every part of the deployment or pod spec can become unwieldy real soon. I agree with issue #50815. We need a mechanism to allow users to customize their deployments, but that shouldn't translate to a flag per customization parameter in the deployment spec. Also, users might want to customize API server and controller manager deployments differently. That ends in having 2 flags for everything in deployment/pod spec and that can go out of control.

Templating the way @quinton-hoole suggested above is the right way to go about it. 

There are two different things being discussed here as @quinton-hoole has pointed out:
1. Being able to pass a file containing kubefed parameters and using that instead of passing individual command line flags. We have discussed this before and I think `ComponentConfig` is the way to go here (https://github.com/kubernetes/kubernetes/issues/40398#issuecomment-276212923).
2. A somewhat related problem is the ability to customize API server and controller manager deployments or services. 
This should be separate from ComponentConfig. For this problem, we should allow users to specify templates for  deployments, services etc.. kubefed can then expand the template, merge the values from command line flags/component config and create those resources. 

Adding individual flags to customize deployment spec is not viable in the long run.

cc @irfanurrehman @shashidharatd @marun

Reply:
@quinton-hoole so do you mean as this, just create a file as following 

```
LiuGuangyas-MacBook-Pro:~ gyliu$ cat args.template
"--bind-address"="0.0.0.0"
"--etcd-servers"="http://localhost:2379"
"--secure-port"=fmt.Sprintf("%d", apiServerSecurePort)
"--client-ca-file"="/etc/federation/apiserver/ca.crt"
"--tls-cert-file"="/etc/federation/apiserver/server.crt"
"--tls-private-key-file"="/etc/federation/apiserver/server.key"
"--admission-control"="NamespaceLifecycle"
```

Then the `kubefed init` will take this file as the source of the arguments?

Reply:
#12245

Reply:
@gyliu513

> @quinton-hoole so do you mean as this, just create a file as following

Actually like `ComponentConfig`. Unfortunately there are no good user docs for them now. This issue #50041 has a few links that show how they look.

> Then the `kubefed init` will take this file as the source of the arguments?

Correct.